File: sdks/dotnet/src/Orvano/OrvanoException.cs
namespace Orvano;

/// <summary>
/// The one exception every Orvano call throws when the server answers with a failure. The server
/// sends RFC 9457 problem details; <see cref="Code"/> is Orvano's stable error code (for example
/// <c>user_already_exists</c>).
/// </summary>
public sealed class OrvanoException : Exception
{
    /// <summary>Creates an <see cref="OrvanoException"/>.</summary>
    /// <param name="status">The HTTP status code.</param>
    /// <param name="code">Orvano's stable error code.</param>
    /// <param name="message">A human readable description of the problem.</param>
    /// <param name="requestId">The request ID, when the server sent one.</param>
    public OrvanoException(int status, string code, string message, string? requestId)
        : base(message)
    {
        Status = status;
        Code = code;
        RequestId = requestId;
    }

    /// <summary>The HTTP status code.</summary>
    public int Status { get; }

    /// <summary>Orvano's stable error code, or <c>unknown</c> when the response carried none.</summary>
    public string Code { get; }

    /// <summary>The request ID to quote when reporting a problem, when the server sent one.</summary>
    public string? RequestId { get; }
}

File: sdks/dart/core/lib/src/orvano_exception.dart
import 'dart:convert';

/// The one exception every Orvano call throws when the server answers with a
/// failure. The server sends RFC 9457 problem details; [code] is Orvano's
/// stable error code (for example `user_already_exists`).
final class OrvanoException implements Exception {
  /// Creates an [OrvanoException].
  const OrvanoException({
    required this.status,
    required this.code,
    required this.message,
    this.requestId,
  });

  /// Reads a failed response body into an [OrvanoException], whatever it
  /// looks like.
  factory OrvanoException.fromResponse(
    int status,
    String body,
    String? requestIdHeader,
  ) {
    Map<String, dynamic> problem = const {};
    try {
      final decoded = jsonDecode(body);
      if (decoded is Map<String, dynamic>) problem = decoded;
    } on FormatException {
      // Not JSON (a proxy error page, for example); fall back to the status.
    }

    String? text(String key) {
      final value = problem[key];
      return value is String && value.isNotEmpty ? value : null;
    }

    return OrvanoException(
      status: status,
      code: text('code') ?? 'unknown',
      message:
          text('detail') ??
          text('title') ??
          'Request failed with status $status',
      requestId: text('requestId') ?? requestIdHeader,
    );
  }

  /// The HTTP status code.
  final int status;

  /// Orvano's stable error code, or `unknown` when the response carried none.
  final String code;

  /// A human readable description of the problem.
  final String message;

  /// The request ID to quote when reporting a problem, when the server sent
  /// one.
  final String? requestId;

  @override
  String toString() => 'OrvanoException($status, $code): $message';
}
